addons/payment_ingenico/models/payment.py

```diff
@@ -151,7 +151,17 @@ def filter_key(key):
         items = sorted((k.upper(), v) for k, v in values.items())
         sign = ''.join('%s=%s%s' % (k, v, key) for k, v in items if v and filter_key(k))
         sign = sign.encode("utf-8")
-        shasign = sha1(sign).hexdigest()
+
+        hash_function = self.env['ir.config_parameter'].sudo().get_param('payment_ingenico.hash_function', 'sha1').lower()
+        if hash_function not in ['sha1', 'sha256', 'sha512']:
+            hash_function = 'sha1'
+
+        if hash_function == 'sha1':
+            shasign = sha1(sign).hexdigest()
+        elif hash_function == 'sha256':
+            shasign = sha256(sign).hexdigest()
+        else:
+            shasign = sha512(sign).hexdigest()
```

Comment:
Since we're whitelisting the hash algorithm, the dispatch should probably be left to `hashlib.new`: it's probably going to do something similar except in straight C (at least when compiled against openssl).

Reply:
Oh I didn't know about hashlib's new